FLT/HIMExperiments/module_topology.lean

```diff
@@ -213,6 +216,111 @@ noncomputable def LinearMap.add (M : Type*) [AddCommGroup M] [Module A M] :
   map_add' _ _ := add_add_add_comm _ _ _ _
   map_smul' _ _ := (DistribSMul.smul_add _ _ _).symm
 
+/-- Basis.repr in the first variable as a linear map. -/
+noncomputable def LinearMap.basis₁ (M : Type*) (ι : Type*) [Finite ι] [AddCommGroup M] [Module A M]
+  (b : Basis ι A M) (i : ι): M →ₗ[A] A where
+  toFun m := b.repr m i
+  map_add' := by simp only [map_add, Finsupp.coe_add, Pi.add_apply, implies_true]
+  map_smul' _ _  := by simp only [LinearMapClass.map_smul, Finsupp.coe_smul, Pi.smul_apply,
+    smul_eq_mul, RingHom.id_apply]
+
+noncomputable def LinearMap.prodfst (M N : Type*) [AddCommGroup M] [Module A M] [AddCommGroup N]
+  [Module A N] : M × N →ₗ[A] M where
+  toFun := Prod.fst
+  map_add' _ _ := rfl
+  map_smul' _ _ := rfl
+
+noncomputable def LinearMap.prodsnd (M N : Type*) [AddCommGroup M] [Module A M] [AddCommGroup N]
+  [Module A N] : M × N →ₗ[A] N where
+  toFun := Prod.snd
+  map_add' _ _ := rfl
+  map_smul' _ _:= rfl
+
+instance Module.instCommAdd {P : Type*} [AddCommGroup P] [Module A P]:
+@ContinuousAdd P (Module.topology A) _ := by
+  apply @ContinuousAdd.mk _ (topology A)
+  rw [Module.prod_canonical A]
+  exact Module.continuous_linear A (LinearMap.add A P)
+
+variable [Module.Finite A M] [Module.Free A M] [Module.Finite A N] [Module.Free A N]
+
+
+-- I really shouldn't start with iA instead of topology A here.
+-- But I need it like this in the next proof and don't know how to rewrite it there
+-- Sorry for leaving it like this...
+instance Module.instContinuousSMul : @ContinuousSMul A M _ _ (topology A) := by
+  let _τM : TopologicalSpace M := Module.topology A
+  apply @ContinuousSMul.mk A M _ _ (topology A)
+  let ι := Free.ChooseBasisIndex A M
+  have b : Basis ι A M := Free.chooseBasis A M
+  suffices Continuous fun (p : A × M) ↦ ∑ i : ι, p.1 • b.repr p.2 i • b i by
+    convert this using 1
+    simp_rw [← Finset.smul_sum, Basis.sum_repr]
+  apply continuous_finset_sum
+  intro i _
+  simp_rw [← mul_smul]
+  suffices Continuous ((fun (a : A) ↦ a • b i) ∘ (fun (m : A × A) ↦ m.1 * m.2)
+    ∘ (fun (m : A × M) ↦ (m.1, b.repr m.2 i))) from this
+  apply Continuous.comp
+  · exact Module.continuous_linear_from_ring A ((LinearMap.lsmul A M).flip (b i))
+  apply Continuous.comp
+  · exact continuous_mul
+  · apply @Continuous.prod_map _ _ _ _ _ _ _ (topology A) (fun m ↦ m) (fun m ↦ b.repr m i)
+    exact continuous_id
+    exact Module.continuous_linear_to_ring A (LinearMap.basis₁ A M ι b i)
```

Comment:
The `continuity` tactic should perhaps be doing that job for you, but maybe it doesn't because it doesn't know about `Module.continuous_linear_to_ring` and there's maybe a way to teach it to the tactic? Or maybe it's hopeless because of the multiple topologies (although you could have declared them earlier in the proof perhaps).

Reply:
I tried `continuity` and it did not do anything.